akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -0,0 +1,325 @@
+        val dependsOn = c.getStringList("depends-on").asScala.toSet
+        val timeout = c.getDuration("timeout", MILLISECONDS).millis
+        val recover = c.getBoolean("recover")
+        k → Phase(dependsOn, timeout, recover)
+      case (k, v) ⇒
+        throw new IllegalArgumentException(s"Expected object value for [$k], got [$v]")
+    }
+  }
+
+  /**
+   * INTERNAL API: https://en.wikipedia.org/wiki/Topological_sorting
+   */
+  private[akka] def topologicalSort(phases: Map[String, Phase]): List[String] = {
+    var result = List.empty[String]
+    var unmarked = phases.keySet ++ phases.values.flatMap(_.dependsOn) // in case phase is not defined as key
+    var tempMark = Set.empty[String] // for detecting cycles
+
+    while (unmarked.nonEmpty) {
+      depthFirstSearch(unmarked.head)
+    }
+
+    def depthFirstSearch(u: String): Unit = {
+      if (tempMark(u))
+        throw new IllegalArgumentException("Cycle detected in graph of phases. It must be a DAG. " +
+          s"phase [$u] depends transitively on itself. All dependencies: $phases")
+      if (unmarked(u)) {
+        tempMark += u
+        phases.get(u) match {
+          case Some(Phase(dependsOn, _, _)) ⇒ dependsOn.foreach(depthFirstSearch)
+          case None                         ⇒
+        }
+        unmarked -= u // permanent mark
+        tempMark -= u
+        result = u :: result
+      }
+    }
+
+    result.reverse
+  }
+
+}
+
+final class CoordinatedShutdown private[akka] (
+  system: ExtendedActorSystem,
+  phases: Map[String, CoordinatedShutdown.Phase]) extends Extension {
+  import CoordinatedShutdown.Phase
+
+  /** INTERNAL API */
+  private[akka] val log = Logging(system, getClass)
+  private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
+  private val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
+  private val tasks = new ConcurrentHashMap[String, Vector[() ⇒ Future[Done]]]
+  private val runStarted = new AtomicBoolean(false)
+  private val runPromise = Promise[Done]()
+
+  private var _jvmHooksLatch = new AtomicReference[CountDownLatch](new CountDownLatch(0))
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] def jvmHooksLatch: CountDownLatch = _jvmHooksLatch.get
+
+  /**
+   * Add a task to a phase. It doesn't remove previously added tasks.
+   * Tasks added to the same phase are executed in parallel without any
+   * ordering assumptions. Next phase will not start until all tasks of
+   * previous phase have been completed.
+   *
+   * Tasks should typically be registered as early as possible after system
+   * startup. When running the coordinated shutdown tasks that have been registered
+   * will be performed but tasks that are added too late will not be run.
+   * It is possible to add a task to a later phase by a task in an earlier phase
+   * and it will be performed.
+   */
+  @tailrec def addTask(phase: String)(task: () ⇒ Future[Done]): Unit = {
```

Comment:
added that and included in log messages, see https://github.com/akka/akka/pull/21930/commits/ba75afd746ffdc4abf31794502d107623611d35a